### ISSUE TITLE
Right way to implement a node with ns index

### ISSUE THREAD

**Issue by mar-ar** [FreeOpcUa/python-opcua#575]
Hi,

I want to implement my own address space with my own namespace. My problem ist that I cant find a way to give a node a namespace. First i thought i can just do it with "ns=" but with the code below my BrowsName is still "0:BaseInterface" and thereby in the standard address space.

`node = ua.AddNodesItem()

node.RequestedNewNodeId = ua.NodeId.from_string("ns=2;i=3")

node.BrowseName = ua.QualifiedName.from_string("BaseInterface")
`

I already found a workaround, but I dont think that this is the best way.
`node.BrowseName = ua.QualifiedName.from_string("2:BaseInterface")`

If someone knows a better way to include the namespace to a node please let me know.

best regards

**Comment by zerox1212** [comment]
Do you really need to work directly with the UA objects? Can you use the high level interface?

**Comment by TryToLearn123** [comment]
No but I thought its the best way to implement a lot of objects, variables and references, like it is used for the standard adressspace.
How would you do this?

**Comment by oroulet** [comment]
What do you want to do? before using the low level api you need to understand the basic opcua concepts. Then you decide if you want to create your nodes using xml or low level api or whatever

**Comment by TryToLearn123** [comment]
I hope that I already have understood the basic opc ua concepts.
I just want to implement my own addresspace with set of objects, variables and references.
And I just saw how it it made with the standard addresspace so I just wanted to make it the same way.

**Comment by zerox1212** [comment]
How many objects do you have? It seems like a lot of work to build nodes with the low level API just to save some startup time.

If you really need to use the low level I can't help you. I haven't found a reason yet to use anything but the nice high level API.

**Comment by TryToLearn123** [comment]
I have somthing around 100 objects. 

What do you mean with the high level API? XML would be also possible for me.

**Comment by zerox1212** [comment]
100 objects or 100 nodes? 100 nodes is nothing...

I mean use the high level API to create your UA nodes:
```python
# get Objects node, this is where we should put our nodes
    objects = server.get_objects_node()

    # populating our address space
    myobj = objects.add_object(idx, "MyObject")
```

**Comment by zerox1212** [comment]
I prefer to use modeler to build an address space design and just import it. For me it's the best organization because then the python code focuses on populating the nodes with data instead of address space design.

**Comment by TryToLearn123** [comment]
My problem therfor is that my addressspace changes in the runtime.

**Comment by oroulet** [comment]
I do not think address space is supposed to change at runtime. You have a design error and will get into trouble

**Comment by TryToLearn123** [comment]
I dont think so. I already saw a few examples where the address space changed while the runtime. Its an advantage of opc ua and the events therefor are defined like GeneralModelcChangeEvent.

**Comment by oroulet** [comment]
I am quite sure this is made for very low frequency changes. By the way the model change event is not implement but this should be easy to implement or send it manually

**Comment by TryToLearn123** [comment]
I think the use case that the information model changes while the runtime is very common.
So there has to be a good way to implement this. And my solution until now was to do it like this.

**Comment by oroulet** [comment]
You can do whatever you want. Even implement an ua server in Photoshop, possible but quite broken by design

**Comment by TryToLearn123** [comment]
I havent found a better way to handle this.
When I start the server I dont know all elements which I want to transfer. One exaple therefor is that I create an error object for each error which occures while the runtime. If more error occure more error objects are created.

**Comment by oroulet** [comment]
I do not know your project but you should consider custom events and/or custom structures.

**Comment by TryToLearn123** [comment]
Its to complex to explain briefly.

What do you mean with custom structures?

**Comment by zerox1212** [comment]
If you don't mind, I'm interested in your use case. What kind of machine would or OPC UA application would be changing it's address space on the fly?

The only thing I can think of would be an HMI development app built on top of OPC UA.

**Comment by TryToLearn123** [comment]
I am implementing an opc ua server for a machine which provides all necessary information about this machine. 
One exaple is like I already mention if the PLC of the machine gets an error I want to read this error (or more) via ua.
An other example is that I provide information about each tool in this machine but the amout of tools changes during time.
Or information about a specific workpiece and so on...

**Comment by oroulet** [comment]
I have seens two cases: 
* a company wanted to expose maintenance shcedule in opcua address space as node. This was wrong. better to have a maintenance variable containing a list of custom Maintenance objects 
* exposing with opcua a simulated shop-floor. the developers sometimes added machines while simulation was running. => This was a good case to modify address space

If you want to expose tools you have two obvious choices.
*  have one node per tools in cnc. Usually there is a fixed number of slots so this model fits perfectly
* have variable node containing a list of tools objects

for error the normal solution is to use events. Send error events. People interested in history can read event history. And maybe update the machine state with current error if this is something important

**Comment by TryToLearn123** [comment]
Hmm thats a good idea. Thanks I have to think about that.

**Comment by zerox1212** [comment]
Don't change the address space for errors by using nodes. Use events and alarms. That way you can enable historizing and clients can read event history.

**Comment by TryToLearn123** [comment]
Yeah sure thanks.

I dont know how to implement a list of objects with properties and components. I know how to realize a variable with an array. Can you explain it please.
